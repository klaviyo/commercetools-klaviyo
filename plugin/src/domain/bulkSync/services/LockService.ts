export interface LockService {
    acquireLock(key: string): void;
    releaseLock(key: string): void;
    checkLock(key: string): void;
}
