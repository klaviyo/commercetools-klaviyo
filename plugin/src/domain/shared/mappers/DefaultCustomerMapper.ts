import { Address, Customer } from '@commercetools/platform-sdk';
import { CustomerMapper } from './CustomerMapper';
import { mapAllowedProperties } from '../../../utils/property-mapper';

const E_164_REGEX = /^\+[1-9]\d{10,14}$/;

export class DefaultCustomerMapper implements CustomerMapper {
    public mapCtCustomerToKlaviyoProfile(customer: Customer, klaviyoProfileId?: string): ProfileRequest {
        const address = this.getCTCustomerAddressForKlaviyo(customer);
        const {
            email,
            firstName: first_name,
            lastName: last_name,
            title,
            companyName: organization,
            id: external_id,
            custom,
        } = customer;
        const props = mapAllowedProperties('customer.customFields', { ...(custom?.fields || {}) });

        return {
            data: {
                type: 'profile',
                id: klaviyoProfileId,
                attributes: {
                    email,
                    external_id,
                    first_name,
                    last_name,
                    title,
                    phone_number: address?.mobile || address?.phone,
                    organization,
                    location: this.mapCTAddressToKlaviyoLocation(address),
                    properties: Object.keys(props).length ? { ...props } : undefined,
                },
            },
        };
    }

    public mapCTAddressToKlaviyoLocation(address?: Address, useSpecialPrefix = false): KlaviyoLocation | null {
        const specialPrefix = useSpecialPrefix ? '$' : '';
        return address
            ? {
                  [`${specialPrefix}address1`]: this.getAddressLine1(address),
                  [`${specialPrefix}address2`]: this.getAddressLine2(address),
                  [`${specialPrefix}city`]: address.city,
                  [`${specialPrefix}country`]: address.country,
                  [`${specialPrefix}region`]: address.region,
                  [`${specialPrefix}zip`]: address.postalCode,
              }
            : null;
    }

    public getAddressLine1(address?: Address): string | null{
        return address
            ? [address.apartment, address.building, address.streetNumber, address.streetName]
                  .filter((element) => element)
                  .join(', ')
            : null;
    }

    public getAddressLine2(address?: Address): string | null {
        return address
            ? [address.additionalStreetInfo, address.additionalAddressInfo].filter((element) => element).join(', ')
            : null;
    }

    // No longer in use, validation only handled by Klaviyo.
    // Phone number removed in worse case scenario.
    public getPhoneNumber(address?: Address): string | null {
        const phoneNumber = address?.mobile || address?.phone;
        if (phoneNumber) {
            const noSpacesAndDashPhoneNumber = phoneNumber.replace(/\s|-|\/|\)|\(|\*|[a-z]|[A-Z]+/g, '');
            return E_164_REGEX.test(noSpacesAndDashPhoneNumber) ? noSpacesAndDashPhoneNumber : null;
        }
        return null;
    }

    /**
     * A CT customer can have multiple addresses, klaviyo accepts only one address information in the location field.
     * This method selects one of the CT addresses to be used to populate the klaviyo location info for a klaviyo profile.
     * The address selection evaluates in order the following conditions and returns the first address found:
     * 1. Selects the customer defaultBillingAddress if set
     * 2. Selects the first available billing address if present
     * 3. Selects the first available address if present
     * @param customer
     */
    public getCTCustomerAddressForKlaviyo(customer: Customer): Address | undefined {
        if (!customer || !customer.addresses.length) {
            return;
        }
        if (customer.defaultBillingAddressId) {
            return customer.addresses?.find((address) => address.id === customer.defaultBillingAddressId);
        }
        if (customer?.billingAddressIds?.length) {
            return customer.addresses?.find((address) => address.id === customer?.billingAddressIds?.find(() => true));
        }
        return customer.addresses?.find(() => true);
    }
}
