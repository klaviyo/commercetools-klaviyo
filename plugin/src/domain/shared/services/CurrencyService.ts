export interface CurrencyService {
  /**
   *
   * @param value the amount to convert (not in cents, e.g. $1.56)
   * @param currencyCode Currency code compliant to [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217).
   */
  convert(value: number, currencyCode: string): number;
}
