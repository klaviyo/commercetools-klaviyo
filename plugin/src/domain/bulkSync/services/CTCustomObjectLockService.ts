import { LockService } from './LockService';
import { ByProjectKeyRequestBuilder } from '@commercetools/platform-sdk/dist/declarations/src/generated/client/by-project-key-request-builder';
import logger from '../../../utils/log';
import { ErrorCodes, StatusError } from '../../../types/errors/StatusError';

export class CTCustomObjectLockService implements LockService {
    private readonly container = 'klaviyo-ct-plugin-lock';
    constructor(private readonly ctApiRoot: ByProjectKeyRequestBuilder) {}
    async acquireLock(key: string): Promise<void> {
        logger.info(`Acquiring lock using CT custom object. Container: ${this.container}, key: ${key}`);
        try {
            await this.ctApiRoot
                .customObjects()
                .withContainerAndKey({
                    container: this.container,
                    key,
                })
                .get()
                .execute();
        } catch (e: any) {
            if (e?.statusCode !== 404) {
                logger.error(`Error getting lock status using CT custom object.`, e);
                throw new StatusError(e.statusCode, e.message, e.code);
            }

            try {
                logger.info(`Lock ${this.container} > ${key} not found. Creating lock using CT custom object`);
                await this.ctApiRoot
                  .customObjects()
                  .post({ body: { container: this.container, key: key, value: '1' } })
                  .execute();
                logger.info(`Lock ${this.container} > ${key} set`)
                return;
            } catch (postError: any) {
                logger.error(`Error getting lock status using CT custom object. ${this.container} > ${key}`, postError);
                throw new StatusError(postError.statusCode, postError.message);
            }

        }
        logger.error(`Lock ${this.container} > ${key} already exists`);
        throw new StatusError(
            409,
            `Lock already exists. Container: ${this.container}, key: ${key}`,
            ErrorCodes.LOCKED,
        );
    }

    async releaseLock(key: string): Promise<void> {
        logger.info(`Releasing lock using CT custom object. Container: ${this.container}, key: ${key}`);
        try {
            await this.ctApiRoot
                .customObjects()
                .withContainerAndKey({
                    container: this.container,
                    key,
                })
                .delete()
                .execute();
            logger.info('Lock released.');
        } catch (e: any) {
            if (e.statusCode !== 404) {
                logger.error(`Error releasing lock using CT custom object. ${this.container} > ${key}`, e);
                throw new StatusError(e.statusCode, e.message);
            }
        }
    }

    async checkLock(key: string): Promise<void> {
        logger.info(`Checking lock using CT custom object. Container: ${this.container}, key: ${key}`);
        try {
            await this.ctApiRoot
                .customObjects()
                .withContainerAndKey({
                    container: this.container,
                    key,
                })
                .get()
                .execute();
        } catch (e: any) {
            if (e?.statusCode !== 404) {
                logger.error(`Error getting lock status using CT custom object.`, e);
                throw new StatusError(e.statusCode, e.message, e.code);
            }
            return;
        }
        logger.error(`Lock ${this.container} > ${key} already exists`);
        throw new StatusError(
            409,
            `Lock already exists. Container: ${this.container}, key: ${key}`,
            ErrorCodes.LOCKED,
        );
    }
}
