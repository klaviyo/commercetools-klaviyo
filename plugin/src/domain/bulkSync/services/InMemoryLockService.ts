import { LockService } from "./LockService";
import { ErrorCodes, StatusError } from "../../../types/errors/StatusError";

/**
 * This is a very simple implementation of the lock service. This solution will not work when multiple instances of this plugin are running.
 * In that case the lock should be stored in a shared location (e.g. file in a bucket, document in a database...).
 * Write your custom implementation of the LockService based on your requirements.
 */
export class InMemoryLockService implements LockService {
    private locks: string[] = [];
    acquireLock(key: string): void {
        this.checkLock(key);
        this.locks.push(key);
    }

    releaseLock(key: string): void {
        const index = this.locks.indexOf(key, 0);
        if (index > -1) {
            this.locks.splice(index, 1);
        }
    }

    checkLock(key:string): void {
        if (this.locks.includes(key)) {
            throw new StatusError(409, 'Cannot acquire lock', ErrorCodes.LOCKED);
        }
    }
}
