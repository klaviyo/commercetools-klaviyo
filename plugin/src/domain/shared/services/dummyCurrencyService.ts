import { CurrencyService } from "./CurrencyService";

export class DummyCurrencyService implements CurrencyService {
    convert(value: number, currencyIso: string): number {
        // ***** hardcoded conversion to USD ***
        // switch (currencyIso) {
        //     case 'GBP':
        //         return value / 1.21;
        //     case 'EUR':
        //         return value / 1.07;
        //     default:
        //         return value;
        // }
        //
        // ***** API service for currencies conversion ***
        // const result = await fetch(`https://my-conversion-api-service.com/${value}?currency=${currencyIso}&targetCurrency=USD`);
        // return result.json();
        //
        // ***** no conversion *****
        return value;
    }
}
