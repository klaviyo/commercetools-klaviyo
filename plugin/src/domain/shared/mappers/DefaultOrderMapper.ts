import { getCustomerProfileFromOrder } from '../../../utils/get-customer-profile-from-order';
import { getTypedMoneyAsNumber } from '../../../utils/get-typed-money-as-number';
import { mapAllowedProperties } from '../../../utils/property-mapper';
import { Category, CategoryReference, LineItem, Order, Product } from '@commercetools/platform-sdk';
import { OrderMapper } from './OrderMapper';
import config from 'config';
import { CurrencyService } from '../services/CurrencyService';
import { CustomerMapper } from './CustomerMapper';
import { getLocalizedStringAsText } from '../../../utils/locale-currency-utils';

export class DefaultOrderMapper implements OrderMapper {
    constructor(private readonly currencyService: CurrencyService, private readonly customerMapper: CustomerMapper) {}
    public mapCtOrderToKlaviyoEvent(
        order: Order,
        orderProducts: Product[],
        metric: string,
        updateAdditionalProfileProperties: boolean,
        time?: string,
    ): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(
                        order,
                        this.customerMapper,
                        updateAdditionalProfileProperties && metric === config.get('order.metrics.placedOrder'),
                    ),
                    metric: {
                        name: metric,
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(order?.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: {
                        ...mapAllowedProperties('order', { ...order }),
                        ...mapAllowedProperties('order.customFields', { ...(order.custom?.fields || {}) }),
                        ItemNames: this.mapOrderLineItemsToItemNames(order),
                        Categories: this.getCategoryNamesFromProduct(
                            orderProducts.map((product) => product.masterData.current.categories).flat(),
                        ),
                    } as any,
                    unique_id: order.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    public mapCtRefundedOrderToKlaviyoEvent(
        order: Order,
        orderProducts: Product[],
        metric: string,
        time?: string,
    ): EventRequest {
        const refundAmounts =
            order.paymentInfo?.payments
                .map((p) =>
                    p.obj?.transactions
                        .filter((t) => t.state === 'Success' && t.type === 'Refund')
                        .map((t) => getTypedMoneyAsNumber(t.amount)),
                )
                .flat()
                .filter((r) => r !== undefined) || [];
        const refundTotal = refundAmounts.length ? refundAmounts.reduce((a, b) => (a || 0) + (b || 0)) || 0 : 0;

        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order, this.customerMapper),
                    metric: {
                        name: metric,
                    },
                    value: this.currencyService.convert(refundTotal, order.totalPrice.currencyCode),
                    properties: {
                        ...mapAllowedProperties('order', { ...order }),
                        ...mapAllowedProperties('order.customFields', { ...(order.custom?.fields || {}) }),
                        ItemNames: this.mapOrderLineItemsToItemNames(order),
                        Categories: this.getCategoryNamesFromProduct(
                            orderProducts.map((product) => product.masterData.current.categories).flat(),
                        ),
                    } as any,
                    unique_id: order.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    public mapOrderLineToProductOrderedEvent(lineItem: LineItem, order: Order, time?: string): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order, this.customerMapper),
                    metric: {
                        name: config.get('order.metrics.orderedProduct'),
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(lineItem.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: { ...lineItem },
                    unique_id: lineItem.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    private mapOrderLineItemsToItemNames(order: Order): string[] {
        const lineItemNames = order.lineItems.map((item) => getLocalizedStringAsText(item.name));
        const customLineItemNames =
            order.customLineItems?.map((item) => getLocalizedStringAsText(item.name)) || [];
        return Array.from(new Set(lineItemNames.concat(customLineItemNames)));
    }

    private getCategoryNamesFromProduct(categories: CategoryReference[]): string[] {
        const categoryNames = categories.map((category) => {
            const categoryAncestorNames = (category.obj as Category).ancestors.map((ancestor) =>
                getLocalizedStringAsText((ancestor.obj as Category).name),
            );
            const categoryName = getLocalizedStringAsText((category.obj as Category).name);
            const discreteCategoryNames = categoryAncestorNames.concat(categoryName);
            return discreteCategoryNames.map((name, index) => {
                const ancestorNames = [];
                for (let i = index; i >= 0; i--) {
                    ancestorNames.push(discreteCategoryNames[i]);
                }
                return ancestorNames.reverse().join(' > ');
            });
        });

        // Conversion from Array -> Set -> Array removes duplicate entries.
        // Useful for certain cases where you have a common main category such as:
        // Women > Pants > Jeans and Women > Seasonal Fashion > Fall
        return Array.from(new Set(categoryNames.flat()));
    }
}
