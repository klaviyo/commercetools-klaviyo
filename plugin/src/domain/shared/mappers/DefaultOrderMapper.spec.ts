import { DefaultOrderMapper } from './DefaultOrderMapper';
import { mock } from 'jest-mock-extended';
import { DummyCurrencyService } from '../services/dummyCurrencyService';
import { sampleOrderCreatedMessage } from '../../../test/testData/orderData';
import { DefaultCustomerMapper } from './DefaultCustomerMapper';

const mockCurrencyService = mock<DummyCurrencyService>();
mockCurrencyService.convert.mockImplementation((value, currency) => value);
const mockCustomerMapper = mock<DefaultCustomerMapper>();
const orderMapper = new DefaultOrderMapper(mockCurrencyService, mockCustomerMapper);
describe('map CT order to Klaviyo event', () => {
    it('should map a commercetools order with a given metric to a klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(
            sampleOrderCreatedMessage.order,
            [],
            'someMetric',
            false,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should map order line items and custom line items to ItemNames property in the klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                lineItems: [
                    {
                        name: {
                            'en-US': 'Product Line Item 1',
                        },
                    },
                ] as any,
                customLineItems: [
                    {
                        name: {
                            'en-US': 'Custom Line Item 1',
                        },
                    },
                ] as any,
            },
            [],
            'someMetric',
            false,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should map product categories to the Categories property in the klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                lineItems: [
                    {
                        name: {
                            'en-US': 'Product Line Item 1',
                        },
                    },
                ] as any,
                customLineItems: [
                    {
                        name: {
                            'en-US': 'Custom Line Item 1',
                        },
                    },
                ] as any,
            },
            [
                {
                    masterData: {
                        current: {
                            categories: {
                                obj: {
                                    name: {
                                        'en-US': 'Test Category 1',
                                    },
                                    ancestors: [],
                                },
                            },
                        },
                    },
                } as any,
            ],
            'someMetric',
            false,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                custom: {
                    type: {
                        typeId: 'type',
                        id: '123456',
                    },
                    fields: {
                        includedField: true,
                        ignoredField: 'test-ignored-field',
                    },
                },
            },
            [],
            'someMetric',
            false,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should map a billing address profile property for a realtime klaviyo event', () => {
        mockCustomerMapper.mapCTAddressToKlaviyoLocation.mockImplementation((address) => {
            const address1 = [address?.apartment, address?.building, address?.streetNumber, address?.streetName]
                  .filter((element) => element)
                  .join(', ');
            return {
                address1,
                city: address?.city,
                country: address?.country,
                region: address?.region,
                zip: address?.postalCode,
            }
        });
        const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                billingAddress: {
                    streetName: 'Tribunali',
                    streetNumber: '14',
                    postalCode: '80100',
                    city: 'Napoli',
                    country: 'IT',
                },
            },
            [],
            'someMetric',
            true,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });
});

describe('map CT refunded order to Klaviyo event', () => {
    it('should map a commercetools refunded order with a given metric to a klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtRefundedOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                paymentInfo: {
                    payments: [
                        {
                            typeId: 'payment',
                            id: '123456',
                            obj: {
                                transactions: [
                                    {
                                        id: '123456',
                                        amount: {
                                            type: 'centPrecision',
                                            currencyCode: 'EUR',
                                            centAmount: 1300,
                                            fractionDigits: 2,
                                        },
                                        type: 'Refund',
                                        state: 'Success',
                                    },
                                ],
                            },
                        },
                    ],
                },
            } as any,
            [],
            'someMetric',
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event', () => {
        const klaviyoEvent = orderMapper.mapCtRefundedOrderToKlaviyoEvent(
            {
                ...sampleOrderCreatedMessage.order,
                custom: {
                    type: {
                        typeId: 'type',
                        id: '123456',
                    },
                    fields: {
                        includedField: true,
                        ignoredField: 'test-ignored-field',
                    },
                },
                paymentInfo: {
                    payments: [
                        {
                            typeId: 'payment',
                            id: '123456',
                            obj: {
                                transactions: [
                                    {
                                        id: '123456',
                                        amount: {
                                            type: 'centPrecision',
                                            currencyCode: 'EUR',
                                            centAmount: 1300,
                                            fractionDigits: 2,
                                        },
                                        type: 'Refund',
                                        state: 'Success',
                                    },
                                ],
                            },
                        },
                    ],
                },
            } as any,
            [],
            'someMetric',
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });
});

describe('map CT order line to klaviyo event', () => {
    it('should map a commercetools order line to a klaviyo Order Product event', () => {
        const klaviyoEvent = orderMapper.mapOrderLineToProductOrderedEvent(
            {
                discountedPricePerQuantity: [],
                lineItemMode: '',
                perMethodTaxRate: [],
                price: {
                    id: 'someId',
                    value: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
                },
                priceMode: '',
                productId: '',
                productType: { id: 'id', typeId: 'product-type' },
                quantity: 0,
                state: [],
                taxedPricePortions: [],
                variant: { id: 1 },
                id: '123-123-123',
                totalPrice: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
                name: {
                    en: 'Test product',
                },
            },
            sampleOrderCreatedMessage.order,
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });
});
