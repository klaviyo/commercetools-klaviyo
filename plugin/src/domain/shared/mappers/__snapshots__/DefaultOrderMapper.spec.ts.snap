// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map CT order line to klaviyo event should map a commercetools order line to a klaviyo Order Product event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Ordered Product",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "discountedPricePerQuantity": [],
        "id": "123-123-123",
        "lineItemMode": "",
        "name": {
          "en": "Test product",
        },
        "perMethodTaxRate": [],
        "price": {
          "id": "someId",
          "value": {
            "centAmount": 1300,
            "currencyCode": "USD",
            "fractionDigits": 2,
            "type": "centPrecision",
          },
        },
        "priceMode": "",
        "productId": "",
        "productType": {
          "id": "id",
          "typeId": "product-type",
        },
        "quantity": 0,
        "state": [],
        "taxedPricePortions": [],
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "variant": {
          "id": 1,
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "123-123-123",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "includedFieldAfterMap": true,
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should map a billing address profile property for a realtime klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should map a commercetools order with a given metric to a klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should map order line items and custom line items to ItemNames property in the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [
          "Product Line Item 1",
          "Custom Line Item 1",
        ],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [
          {
            "name": {
              "en-US": "Custom Line Item 1",
            },
          },
        ],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [
          {
            "name": {
              "en-US": "Product Line Item 1",
            },
          },
        ],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should map product categories to the Categories property in the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [
          "Test Category 1",
        ],
        "ItemNames": [
          "Product Line Item 1",
          "Custom Line Item 1",
        ],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [
          {
            "name": {
              "en-US": "Custom Line Item 1",
            },
          },
        ],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [
          {
            "name": {
              "en-US": "Product Line Item 1",
            },
          },
        ],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT refunded order to Klaviyo event should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "includedFieldAfterMap": true,
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT refunded order to Klaviyo event should map a commercetools refunded order with a given metric to a klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "Categories": [],
        "ItemNames": [],
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;
